File: py/aoc2023/day5.py
"""
Day 5:
"""

import re
from functools import reduce
from operator import itemgetter

SAMPLE_INPUT = """
seeds: 79 14 55 13

seed-to-soil map:
50 98 2
52 50 48

soil-to-fertilizer map:
0 15 37
37 52 2
39 0 15

fertilizer-to-water map:
49 53 8
0 11 42
42 0 7
57 7 4

water-to-light map:
88 18 7
18 25 70

light-to-temperature map:
45 77 23
81 45 19
68 64 13

temperature-to-humidity map:
0 69 1
1 0 69

humidity-to-location map:
60 56 37
56 93 4
"""

NUMBER_RE = re.compile(r"\d+")


def _parse(data):
    stanzas = data.split("\n\n")
    return [int(m.group(0)) for m in NUMBER_RE.finditer(stanzas[0])], [
        sorted(
            (
                (nums[1], nums[1] + nums[2], nums[0] - nums[1])
                for line in stanza.splitlines()
                if len(nums := [int(m.group(0)) for m in NUMBER_RE.finditer(line)]) == 3
            ),
            key=itemgetter(0),
        )
        for stanza in stanzas[1:]
    ]


def _remap(ranges, mappings):
    for start, end in ranges:
        for start2, end2, offset in mappings:
            if start2 >= end or start >= end2:
                continue
            if start < start2:
                yield start, start2
                start = start2
            end2 = min(end, end2)
            yield start + offset, end2 + offset
            start = end2
        if start < end:
            yield start, end


def part1(data):
    """
    >>> part1(SAMPLE_INPUT)
    35
    """
    seeds, mappings = _parse(data)
    return min(
        map(itemgetter(0), reduce(_remap, mappings, ((x, x + 1) for x in seeds)))
    )


def part2(data):
    """
    >>> part2(SAMPLE_INPUT)
    46
    """
    seeds, mappings = _parse(data)
    return min(
        map(
            itemgetter(0),
            reduce(_remap, mappings, ((x, x + y) for x, y in zip(*[iter(seeds)] * 2))),
        )
    )


parts = (part1, part2)

File: rs/src/day5.rs
use itertools::Itertools;

struct Mapping {
    start: i64,
    end: i64,
    offset: i64,
}

fn parse(data: &str) -> (Vec<i64>, Vec<Vec<Mapping>>) {
    let Some((first, rest)) = data.split_once("\n\n") else {
        return (vec![], vec![]);
    };
    (
        first
            .split(|c: char| !c.is_ascii_digit())
            .filter_map(|s| s.parse::<i64>().ok())
            .collect(),
        rest.split_terminator("\n\n")
            .map(|stanza| {
                stanza
                    .lines()
                    .filter_map(|line| {
                        let mut iter = line.splitn(3, ' ');
                        let dest = iter.next()?.parse::<i64>().ok()?;
                        let source = iter.next()?.parse::<i64>().ok()?;
                        let length = iter.next()?.parse::<i64>().ok()?;
                        Some(Mapping {
                            start: source,
                            end: source + length,
                            offset: dest - source,
                        })
                    })
                    .sorted_by_key(|mapping| mapping.start)
                    .collect()
            })
            .collect(),
    )
}

fn remap(mappings: &[Mapping], range: (i64, i64)) -> Vec<(i64, i64)> {
    let mut ranges = Vec::new();
    let acc = mappings
        .iter()
        .filter(|mapping| mapping.start < range.1 && range.0 < mapping.end)
        .fold(range.0, |acc, mapping| {
            let start = acc.max(mapping.start);
            let end = range.1.min(mapping.end);
            if acc < start {
                ranges.push((acc, start));
            }
            ranges.push((start + mapping.offset, end + mapping.offset));
            end
        });
    if acc < range.1 {
        ranges.push((acc, range.1));
    }
    ranges
}

pub fn part1(data: &str) -> Option<i64> {
    let (seeds, mappings) = parse(data);
    mappings
        .into_iter()
        .fold(
            seeds.into_iter().map(|x| (x, x + 1)).collect::<Vec<_>>(),
            |ranges, mappings| {
                ranges
                    .into_iter()
                    .flat_map(|range| remap(&mappings, range))
                    .collect()
            },
        )
        .into_iter()
        .map(|(x, _)| x)
        .min()
}

pub fn part2(data: &str) -> Option<i64> {
    let (seeds, mappings) = parse(data);
    mappings
        .into_iter()
        .fold(
            seeds
                .into_iter()
                .chunks(2)
                .into_iter()
                .filter_map(|chunk| {
                    let mut iter = chunk.into_iter();
                    let x = iter.next()?;
                    let y = iter.next()?;
                    Some((x, x + y))
                })
                .collect::<Vec<_>>(),
            |ranges, mappings| {
                ranges
                    .into_iter()
                    .flat_map(|range| remap(&mappings, range))
                    .collect()
            },
        )
        .into_iter()
        .map(|(x, _)| x)
        .min()
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;
    use pretty_assertions::assert_eq;

    static EXAMPLE: &str = indoc! {"
        seeds: 79 14 55 13

        seed-to-soil map:
        50 98 2
        52 50 48

        soil-to-fertilizer map:
        0 15 37
        37 52 2
        39 0 15

        fertilizer-to-water map:
        49 53 8
        0 11 42
        42 0 7
        57 7 4

        water-to-light map:
        88 18 7
        18 25 70

        light-to-temperature map:
        45 77 23
        81 45 19
        68 64 13

        temperature-to-humidity map:
        0 69 1
        1 0 69

        humidity-to-location map:
        60 56 37
        56 93 4
    "};

    #[test]
    fn part1_examples() {
        assert_eq!(Some(35), part1(EXAMPLE));
    }

    #[test]
    fn part2_examples() {
        assert_eq!(Some(46), part2(EXAMPLE));
    }
}
